### PR TITLE
[AArch64] Expand vector ops when NEON and SVE are unavailable.

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
Unlike `+noneon` we must assume that vector types are available, i.e. it is
valid to pass/return vector arguments to and from functions. However, the
compiler must make sure to scalarize any vector operations.

### CODE REVIEW

**Comment by sdesmalen-arm** [comment]
This PR is currently still based off #90723 but I'll rebase once that PR lands.

The patch to review is: https://github.com/llvm/llvm-project/pull/90833/commits/2270ec14ae56733c0b65064c109f2ece6215829f

**Review by kmclaughlin-arm — APPROVED** [review]

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -378,6 +378,27 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
     addQRTypeForNEON(MVT::v2i64);
     addQRTypeForNEON(MVT::v8f16);
     addQRTypeForNEON(MVT::v8bf16);
+  } else if (Subtarget->hasNEON() || Subtarget->useSVEForFixedLengthVectors()) {
+    addRegisterClass(MVT::v16i8, &AArch64::FPR8RegClass);
+    addRegisterClass(MVT::v8i16, &AArch64::FPR16RegClass);
+
+    addRegisterClass(MVT::v2f32, &AArch64::FPR64RegClass);
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:385]: I don't really like the duplication here.  What about passing `isNeonAvailable()` into `addDRTypeForNEON` and `addQRTypeForNEON` to toggle the call to `addTypeForNEON`?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -1328,6 +1349,24 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
     // FADDP custom lowering
     for (MVT VT : { MVT::v16f16, MVT::v8f32, MVT::v4f64 })
       setOperationAction(ISD::FADD, VT, Custom);
+  } else {
```
[the remark this answers is not in the document]
Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:1352]: Alternatively should this be `else if (!Subtarget->useSVEForFixedLengthVectors)`? Just thinking there's several instances of "set everything to Expand" going on.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 1 above)
Comment (sdesmalen-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: I've moved this into (the renamed functions) `addQRType` and `addDRType` now.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp [hunk 3]
```diff
@@ -3731,8 +3731,10 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
   }
   case ISD::SUB: {
     EVT VT = Node->getValueType(0);
-    assert(TLI.isOperationLegalOrCustom(ISD::ADD, VT) &&
-           TLI.isOperationLegalOrCustom(ISD::XOR, VT) &&
+    assert((VT.isFixedLengthVector() || // fixed length ADD can be expanded to
```
Comment (paulwalker-arm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:3734]: This change looks weird.  The assert is checking whether it is safe to create `ADD` and `XOR` nodes, which you're bypassing?  What does the failing test case look like?  Part of me thinks the original assert should be an `if` statement, but then I'm also wondering why `VectorLegalizer::LegalizeOp` hasn't already unrolled the operation.

Comment (sdesmalen-arm) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:3734]: That's right. The way the code now works (if you'd ignore the assert) is that it expands a `sub` into a `add` + `xor`. If those are vector operations that also require expansion, they can be expanded further (in a subsequent step) into scalar operations. The assert is trying to match the way the code currently works, which allows the code to be successfully compiled. I figured updating the assert was the most sensible thing to do for now, given that it's not really a common use-case to optimise for.

Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:3734]: To me the assert exists to ensure optimal code, which is to say it doesn't really make sense to convert the sub to an inverted add unless both those operations are in someway legal.  When not the better result would be to just expand the sub.  Which I guess makes the assert a bear trap for the first person that needs the support.

**Review by paulwalker-arm — APPROVED** [review]
Given the size of the generated code I think we might start to question the usefulness of some of the tests (e.g. testing v32i8 variants).

**Pull request merged**